packages/cli/src/commands/hydrogen/deploy.ts

```diff
@@ -112,47 +138,98 @@ export default class Deploy extends Command {
 
 interface OxygenDeploymentOptions {
   environmentTag?: string;
+  force: boolean;
+  noJsonOutput: boolean;
   path: string;
   publicDeployment: boolean;
   shop: string;
   token?: string;
+  metadataDescription?: string;
   metadataUrl?: string;
   metadataUser?: string;
   metadataVersion?: string;
 }
 
 interface GitCommit {
   refs: string;
+  hash: string;
 }
 
 export async function oxygenDeploy(
   options: OxygenDeploymentOptions,
 ): Promise<void> {
   const {
     environmentTag,
+    force: forceOnUncommitedChanges,
+    noJsonOutput,
     path,
     shop,
     publicDeployment,
     metadataUrl,
     metadataUser,
     metadataVersion,
   } = options;
-  const ci = ciPlatform();
+  let {metadataDescription} = options;
+
+  let isCleanGit = true;
+  try {
+    await ensureIsClean(path);
+  } catch (error) {
+    if (error instanceof GitDirectoryNotCleanError) {
+      isCleanGit = false;
+    }
+    if (!forceOnUncommitedChanges && !isCleanGit) {
+      renderWarning({
+        body: 'Uncommitted changes detected.',
+        nextSteps: [
+          [
+            'Commit your changes before deploying or use the ',
+            {command: '--force'},
+            ' flag to deploy with uncommitted changes.',
+          ],
+        ],
+      });
+      return;
+    }
+  }
+
+  const isCI = ciPlatform().isCI;
   let token = options.token;
   let branch: string | undefined;
+  let commitHash: string | undefined;
   let deploymentData: OxygenDeploymentData | undefined;
   let deploymentEnvironmentTag: string | undefined = undefined;
   let gitCommit: GitCommit;
 
   try {
     gitCommit = await getLatestGitCommit(path);
     branch = (/HEAD -> ([^,]*)/.exec(gitCommit.refs) || [])[1];
+    commitHash = gitCommit.hash;
   } catch (error) {
     outputWarn('Could not retrieve Git history.');
     branch = undefined;
   }
 
-  if (!ci.isCI) {
+  if (!metadataDescription && !isCleanGit) {
+    renderWarning({
+      body: 'Deploying uncommited changes, but no description has been provided.',
+      nextSteps: [
+        [
+          'Use the ',
+          {command: '--metadata-description'},
+          ' flag to provide a description.',
+        ],
+        [
+          'If no description is provided, the description defaults to ',
+          {userInput: '<sha> with additional changes'},
+          ' using the SHA of the last commit.',
+        ],
+      ],
+    });
+    metadataDescription = `${commitHash} with additional changes`;
```

Comment:
I find it a bit confusing that we show "nextSteps" but we are continuing the execution anyway. Maybe it would be better to just show a unique warning with `Deploying uncommited changes, but no description has been provided with --meta-description flag. Defaulting to  XYZ.` or similar?